### ISSUE TITLE
Create an interface for Kover tasks

### ISSUE THREAD

**Issue by mgroth0** [Kotlin/kotlinx-kover#338]
I am new to this plugin and I wanted to make certain tasks depend on Kover tasks. However, `KoverHtmlTask` and such are `internal`. This would be fine as long a superclass or interface were made public. But even `AbstractKoverReportTask` is internal. There is no straightforward way to do `task.dependsOn(tasks.withType<KoverHtmlTask>())`. 

There are a couple of simple solutions. One is that you can make these tasks public and make the constructor internal. Another is that you can create a public interface that they extend from.

**Comment by shanshin** [comment]
Hi,
it is not worth using task types for dependencies, because for example, in Android, several tasks with the same type can be created (in the future this may be added to regular JVM projects).

You may add a dependency to a Kover task by its name, for example
`tasks.myTask { dependsOn(tasks.named("koverHtmlReport")) }`

Names `koverXmlReport`, `koverHtmlReport`, `koverVerify` are stable for Kotlin/JVM or Kotlin/MPP, сonstants for the names of these tasks will be introduced in `0.7.0`.

**Comment by shanshin** [comment]
Your problem is solved by searching for a task by name (`tasks.named("someName")`), while it is also safe because if there is no task with that name, an error will be thrown.

The use of `tasks.withType<X>.configureEach` in the case of Kover tasks in most cases is not assumed, because all the configuration of the plug-in takes place in a single place - extensions of the project.

For rare cases, a search by name can be used, which is generally preferable, because in Android projects most often work will be carried out with a specific build variant and the corresponding task.

>  I've never encountered a gradle plugin before this that doesn't have public task classes, regardless how many tasks are created per project.

In fact, there are quite a lot of such tasks, for example, the most frequently used such task is `check` from the `base` plugin.

**Comment by shanshin** [comment]
In the future, some public classes may be added, however, it is necessary to understand what problem they will solve.

At the moment, the only known problem is defining dependencies between tasks, but it is easily solved with the names. For kts build scripts, we will add type-safe accessors for default reports so that users can write `tasks.koverHtmlReport`.

**Comment by qwwdfsad** [comment]
>Given how small of a change this is that there are identifiable benefits, I'm curious why there is such a need for justification. What costs are there to simply removing the internal keyword?

It immediately becomes a part of public API which imposes quite a lot of restrictions:

* API members should be properly documented
* The API itself should be named properly, be consistent, and hack-free. E.g. now we can get away with `val someTrickFlagAsWorkaround`, but with public API we can't
* It's part of API, thus extra care is required regarding API evolution. 

Note that all these problems are solvable and **are not** arguments against your proposal, but rather an answer to "why it's not public by default" question.

To the original report -- thanks for the effort and for going the extra mile in explaining to us why this is a problem for Kover's users.
We'll discuss it internally and see what we can do to both solve your issue and keep our public API minimal

**Comment by shanshin** [comment]
> tasks.named("someName") is less safe because it only throws an error at runtime, not compilation time.

This problem is solved by using constants with the task names: e.g. `kotlinx.kover.api.KoverNames.HTML_REPORT_TASK_NAME` for `0.6.1`

> What issues would be introduced by simply removing the internal keyword for the existing task classes?

Unfortunately in this case `AbstractKoverReportTask` becomes public, however these are implementation details. This class does not fit the hierarchy of reports described above (verify is also inherited from it), however, by making it public, we will need to maintain exactly such a hierarchy in the future for compatibility, which makes refactoring difficult.

 > Finding objects by type rather than name is also useful because if the hierarchical structure.

At this point, there may be discrepancies in the understanding of the correct hierarchy. After the implementation of #229, new tasks for printing "reports to logs" will appear, and verification will start working in a similar way after #339.

> Allowing users to use tasks.withType does not prevent the from using tasks.named. tasks.named, kotlin-gradle-dsl accessors, or other techniques can still be used to find tasks for a particular android build varient or similar.

in this case, Android users can find an example of a solution with `withType` on the Internet and use `configureEach` to configure several tasks at once without realizing it.


If the concepts of default reports (always existing tasks with names `koverHtmlReport`, `koverXmlReport` etc) remain in the stable version , then it will be possible to add the interfaces like `KoverDefaultHtmlReport`. However, this requires additional design in the future.

**Comment by aSemy** [comment]
I agree that making the tasks `internal` is highly unusual, and in the case of Groovy build scripts has no effect, since Groovy can ignore visibility modifiers.

```groovy
// build.gradle

import kotlinx.kover.gradle.plugin.tasks.reports.KoverHtmlTask;

plugins {
  id "org.jetbrains.kotlin.jvm" version "1.8.21"
  id "org.jetbrains.kotlinx.kover" version "0.7.1"
}

afterEvaluate { // required because of https://github.com/Kotlin/kotlinx-kover/issues/394
  tasks.named("koverHtmlReport", KoverHtmlTask).map {
    it.reportDir
  }
}
```

Another usecase for exposing the task types is to be able to access the input/output properties in a typesafe way. For example, I might want to use the files that a Kover task generates as the input for another task https://docs.gradle.org/8.1.1/userguide/incremental_build.html#sec:task_input_output_side_effects. Because the classes are private then I can't do this.

**Comment by shanshin** [comment]
> For example, I might want to use the files that a Kover task generates as the input for another task https://docs.gradle.org/8.1.1/userguide/incremental_build.html#sec:task_input_output_side_effects. Because the classes are private then I can't do this.

You may explicitly specify the report output directory and add a `dependsOn("someKoverTask")` - this will be enough for Gradle.

Until the DSL stabilizes, this will be more efficient than getting migration errors when changing the implementation of Kover tasks classes.

**Comment by aSemy** [comment]
You're right, I shouldn't have said "I can't do this". 

However, because the classes are private it makes it more difficult to discover this behaviour without digging through the source code or being very savvy with Gradle. It also makes it more difficult to perform actions on a tasks output, e.g. filtering the output

```kotlin
val someSpecificKoverFiles = tasks.someKoverTask.map { 
  it.reportDir.asFileTree.matching { include(...) }.files 
}
```

Also, as you have indicated you might refactor the task - so are you sure it's safe to use `dependsOn("someKoverTask")`? The task name or output could change without notice.

**Comment by shanshin** [comment]
> and in the case of Groovy build scripts has no effect, since Groovy can ignore visibility modifiers.

Java reflection is also available, but the question here is not to exclude the technical possibility of accessing the class, but to ensure that users do not randomly start using implementation details for which we do not guarantee backward compatibility.


> It also makes it more difficult to perform actions on a tasks output, e.g. filtering the output

This is a good use case, the outputs of the task can be listed in the public interface, but the inputs should not be specified when the configuration is carried out through the project extension.

> Also, as you have indicated you might refactor the task - so are you sure it's safe to use dependsOn("someKoverTask")? The task name or output could change without notice.

This approach is still under development, but it seems to me that constants should be declared for Kover default tasks names, and for Android tasks it is necessary to create a public function like `koverHtmlreportTaskName(buildVariantName)`

**Comment by aSemy** [comment]
> This is a good use case, the outputs of the task can be listed in the public interface, but the inputs should not be specified when the configuration is carried out through the project extension.

I strongly disagree that Kover has to be so defensive with regards to protecting task properties. Manually modifying the inputs is sometimes necessary. To give one example: [to help with integration tests](https://github.com/adamko-dev/dokkatoo/blob/a4d0987a70f8955cfa4bfe940c50ba3812e761bf/modules/dokkatoo-plugin-integration-tests/projects/it-android-0/dokkatoo/build.gradle.kts#L18-L31).

These properties are already exposed to the public in Groovy scripts, so keeping the tasks as private just makes `gradle.kts` scripts inferior. 

I disagree that Java reflection to access private code is comparable to Groovy - the former is complicated and the latter is not easily accessible, and is even suggested via auto-completion.

The approach at the moment is _encouraging_ using random implementation details. Because of using stringly-typed task names, inferring task outputs, and `afterEvaluate {}` [right now I'm writing non-type-safe scripts](https://github.com/ethauvin/urlencoder/blob/8bf5bb8349defc38fd411e7a6bd0909dbdfd04cf/buildSrc/src/main/kotlin/buildsrc/conventions/code-quality.gradle.kts#L42-L44) that will degrade over time, and are missing out on unbelievably helpful functionality like `@Deprecate`, [`ReplaceWith`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-replace-with/), and [`@ReplacedBy`](https://docs.gradle.org/current/javadoc/org/gradle/api/model/ReplacedBy.html).

> This approach is still under development, but it seems to me that constants should be declared for Kover default tasks names, and for Android tasks it is necessary to create a public function like `koverHtmlreportTaskName(buildVariantName)`

I understand that dealing with Android is complicated, but thought of having to deal with a plugin that is breaking the norms that all other plugins follow is exhausting. Using Gradle is already difficult as it is. I shouldn't have to leave my IDE in order to figure out how Kover wants me to use Gradle.

**Comment by cldfzn** [comment]
In addition, when writing custom plugins and reacting to other plugins being enabled. Having to guess the task name is not an effective solution. Being able to say "if this task is present in the task graph, finalize it with this task", so that a custom plugin can post process the report generated. With the current behavior I have to duplicate the logic this plugin uses to generate task names, so I can iterate through them to create the necessary dependencies and finalizers.

**Comment by shanshin** [comment]
Implemented in `0.8.0-Beta`.
This is not a production ready release, the plugin DSL is experimental and can be changed after receiving feedback.

**Comment by shanshin** [comment]
Make sense to extend `org.gradle.api.Task` in `KoverReport`.

**Comment by shanshin** [comment]
Could you clarify for what purpose `TaskProvider` may be needed, and the provider of which task?

**Comment by shanshin** [comment]
For Kotlin JVM project you already can use `tasks.koverVerify`, `tasks.koverHtmlReport` etc.

For Android projects we can add extensions like `fun TaskContainer.koverVerify(variantName: String)` to use it this way `tasks.koverVerify("debug")` (applicable for `kts` build scripts).

**Comment by shanshin** [comment]
> This would work in the same way as org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget#binaries (and many other similar features of the kotlin gradle plugin)?

Comparing with the KMP plugin, you miss a small detail that there is both the creation of a target and access to its provider in one place.

E.g. for example
```
kotlin {
    macosArm64 {
        binaries.configureEach {
            someUsage(linkTaskProvider)
        }
    }
}
```
you take the task in the extension in which you actually create it.

When it comes to Kover + AGP, the build variants are specified in AGP, in Kover we cannot know in advance which variants exist and which do not (in some cases they may be created at the `afterEvaluate` stage).

Accordingly, even if, for example, we provide such a DSL
```
kover {
    reports {
        variant("debug") {
            htmlReportProvider: TaskProvider<KoverHtmlReport>
        }
    }
}
```
then it is necessary to write a custom implementation of `TaskProvider`, because the task will be created much later than the access to the provider will take place.

---

As a compromise option, I can suggest adding the `variantName` property to the `KoverReport` type, so that you can write such configurations
```
val koverVerifyTasks = tasks.matching {
    it is KoverVerifyReport && it.variantName in setOf("debug", "release")
}

myTask.configure { 
    dependsOn(koverVerifyTasks)
}
```

**Comment by shanshin** [comment]
Yes, `matching` has some problems, but the effect on performance will be negligible because in the example I mentioned, `withType` is used as in the Gradle recommendation.
Since in your case you wanted to take almost all variants except JVM, then a minimum of unnecessary tasks will be created (for example, only for `jvm` and total)

**Comment by shanshin** [comment]
It's about current needs.
Supporting all multiplatform targets is a huge task that will take a very long time to develop. It is not known what DSL will be then, and whether it will be a separate Gradle plugin. 
Therefore, it is worth not to go so far.

**Comment by shanshin** [comment]
Thanks!

I will still add `variantName` attribute to the `KoverReport` interface.
This may not be an ideal option, but right now there is no other way to filter Kover tasks other than analyzing the task name and this is being implemented quickly.

**Comment by shanshin** [comment]
Fixed in `0.8.0-Beta2`

**Comment by shanshin** [comment]
Relates #587
